capsules/src/sha256.rs

```diff
@@ -0,0 +1,402 @@
+        message_schedule[14] = (length64 >> 32) as u32;
+        message_schedule[15] = (length64 & 0xffffffff) as u32;
+        self.perform_sha(&mut message_schedule);
+        buf.slice(len..len);
+    }
+
+    fn perform_sha(&self, message_schedule: &mut [u32; 64]) {
+        // Message schedule
+        for i in 16..64 {
+            let mut s0 = self.right_rotate(message_schedule[i - 15], 7);
+            s0 ^= self.right_rotate(message_schedule[i - 15], 18);
+            s0 ^= message_schedule[i - 15] >> 3;
+            let mut s1 = self.right_rotate(message_schedule[i - 2], 17);
+            s1 ^= self.right_rotate(message_schedule[i - 2], 19);
+            s1 ^= message_schedule[i - 2] >> 10;
+            message_schedule[i] = message_schedule[i - 16] + s0 + message_schedule[i - 7] + s1;
+        }
+
+        // Compression
+        let mut hashes = self.hash_values.get();
+        for i in 0..64 {
+            let s1 = self.right_rotate(hashes[4], 6)
+                ^ self.right_rotate(hashes[4], 11)
+                ^ self.right_rotate(hashes[4], 25);
+            let ch = (hashes[4] & hashes[5]) ^ ((!hashes[4]) & hashes[6]);
+            let constant = self.round_constants.map_or(0, |k| k[i]);
+            let temp1 = hashes[7] + s1 + ch + constant + message_schedule[i];
+            let s0 = self.right_rotate(hashes[0], 2)
+                ^ self.right_rotate(hashes[0], 13)
+                ^ self.right_rotate(hashes[0], 22);
+            let maj = (hashes[0] & hashes[1]) ^ (hashes[0] & hashes[2]) ^ (hashes[1] & hashes[2]);
+            let temp2 = s0 + maj;
+
+            hashes[7] = hashes[6];
+            hashes[6] = hashes[5];
+            hashes[5] = hashes[4];
+            hashes[4] = hashes[3].wrapping_add(temp1);
+            hashes[3] = hashes[2];
+            hashes[2] = hashes[1];
+            hashes[1] = hashes[0];
+            hashes[0] = temp1.wrapping_add(temp2);
+        }
+
+        let mut new_hashes = self.hash_values.get();
+        for i in 0..8 {
+            new_hashes[i] = new_hashes[i].wrapping_add(hashes[i]);
+        }
+        self.hash_values.set(new_hashes);
+    }
+}
+
+impl<'a> DigestData<'a, 32> for Sha256Software<'a> {
+    fn add_data(
+        &self,
+        data: LeasableBuffer<'static, u8>,
+    ) -> Result<usize, (ErrorCode, &'static mut [u8])> {
+        if self.busy() {
+            Err((ErrorCode::BUSY, data.take()))
+        } else {
+            self.state.set(State::Data);
+            if self.handle.is_none() {
+                Err((ErrorCode::FAIL, data.take()))
+            } else {
+                let len = data.len();
+                self.handle.map(|handle| {
+                    self.deferred_caller.set(*handle);
+                    self.input_data.set(data);
+                    self.compute_sha256();
+                });
+                Ok(len)
+            }
+        }
+    }
+
+    fn clear_data(&self) {
+        let _ = self.initialize();
```

Comment:
This won't clear the data if we are busy. The lack of errors indicates that this should clear the state even if the implementation is busy

Reply:
The issue I saw here is there's no return value, so its behavior needs to be deterministic. This will clear the data if it's busy? This method can't be called in the middle of an operation. It will possibly clear data that has been filled in through a SHA but whose callback is pending through the deferred call. The `DigestData` HIL doesn't say what should happen here. So I went for clearing what's in state.

Reply:
The HIL should probably have a `Result` here and say what happens.

Reply:
If the callback is pending then we should clear the data and set `State::Idle`. The idea with `clear_data()` is that it can't fail, we need the data cleared when we call it

Reply:
What about the callback? I think we want to specify this in terms of the behavior as a caller sees it (design) rather than what the capsule does (implementation). So, in particular, if there are operations pending what happens to their callbacks? What error values do they return?

You can't move to the idle state because you might leak memory. E.g., 
```rust
sha.verify(verify_buf);
sha.clear_data(); 
sha.verify(verify_buf2); // verify_buf lost here?
```

Reply:
After `clear_data()` the implementation should trigger a callback with the `ErrorCode::CANCEL` to return the buffer

Reply:
> After `clear_data()` the implementation should trigger a callback with the `ErrorCode::CANCEL` to return the buffer

This doesn't quite agree with TRD2. 

https://github.com/tock/tock/blob/master/doc/reference/trd2-hil-design.md#rule-3-split-phase-result-values-indicate-whether-a-callback-will-occur

The point I'm getting at is that `clear_data` needs to return something. It sounds like you are arguing that case 1 (not cancelled) should not occur. The example here would be that a client calls `hash`, the implemention completes the hash and pushes a deferred procedure call with the completed result, then the client calls `clear_data`. In this case, the hash would just be thrown away and `hash_done` would have an ECANCEL? In this case I think that's feasible (there are others when it's not, e.g., communication, you want to signal that the communication was in fact sent).

Reply:
The idea behind not returning an error is that `clear_data()` can't fail. Let's say the application crashes, it wants to call `clear_data()` to clear any secrets and know that it has been cleared. So yes, the not cancelled case should not occur.

Exactly, in that case `hash_done()` should be called with `ECANCEL`.

The current lowRISC implementation doesn't do that (it just wipes everything), so we loose the buffer (that's a bug) but it doesn't leak information by not cancelling the operation.

If we do have a good reason to allow cancels to fail, then the HIL needs to be modified to reflect that